### PR TITLE
SSCSSI-331 | Add initial delays and maxdelay for async corresponding processing

### PR DESCRIPTION
**Pull request by adusumillipraveen**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/SSCSSI-331 

### Change description ###

- Setting maxDelay as the default is 30 seconds if unspecified thus limiting the overall time for retries to just under 2 minutes.
- Nearly 7%  of the PDFs are not ready by then, tuning maxRetries, delays and multiplier to wait for a maximum of15 minutes.
- Also adding initialDelay for first attempt with Thread.sleep as first attempt is always failing. 

- Alternative approach to handle Thread.Sleep will be something like, but we cannot avoid the Thread from waiting to continue further.

```Java
  if(context != null && context.getRetryCount() == 0){
        log.info("adding additional wait");
        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
        try {
        CompletableFuture<Void> delayedFuture = CompletableFuture.runAsync(() -> {}, CompletableFuture.delayedExecutor(initialDelay, TimeUnit.MILLISECONDS, executor));
        delayedFuture.join();
        } finally {
            executor.shutdown();
        }
    }
```

- or a more complex approach like below as exception handling is key for Retryable to work as expected.

 ```Java
ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();

        try {

            // Use CompletableFuture to introduce a delay before throwing the exception
            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {}, executor)
                .thenRunAsync(() -> {
                    try {
                        final byte[] pdfForLetter = client.getPdfForLetter(notificationId);
                        log.info("Using merge letter correspondence V2 to upload letter correspondence for {} ", ccdCaseId);
                        ccdNotificationsPdfService.mergeLetterCorrespondenceIntoCcdV2(pdfForLetter, Long.valueOf(ccdCaseId), correspondence);
                    } catch (NotificationClientException e) {
                        throw new CompletionException(e);
                    }
                }, CompletableFuture.delayedExecutor(5, TimeUnit.SECONDS, executor));

            // Wait for the future to complete to catch the exception
            future.join();
        } catch (CompletionException e) {
            Throwable cause = e.getCause();
            if (cause instanceof NotificationClientException) {
                NotificationClientException nce = (NotificationClientException) cause;
                if (nce.getMessage().contains("PDFNotReadyError")) {
                    log.info("Got a PDFNotReadyError back from gov.notify for case id: {}.", ccdCaseId);
                } else {
                    log.warn("Got a strange error '{}' back from gov.notify for case id: {}.", nce.getMessage(), ccdCaseId);
                }
            } else {
                log.error("An error occurred", e);
            }
            throw e;
        } finally {
            // Shutdown the executor
            executor.shutdown();
        }
    }
```

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java [hunk 1]
```diff
@@ -43,13 +60,7 @@ public void saveLetter(NotificationClient client, String notificationId, Corresp
...
 
     @Async
-    @Retryable(maxAttemptsExpression = "#{${letterAsync.maxAttempts}}", backoff = @Backoff(delayExpression = "#{${letterAsync.delay}}", multiplierExpression = "#{${letterAsync.multiplier}}", random = true))
-    public void saveLetter(Correspondence correspondence, final byte[] pdfForLetter, String ccdCaseId) {
-        ccdNotificationsPdfService.mergeLetterCorrespondenceIntoCcd(pdfForLetter, Long.valueOf(ccdCaseId), correspondence);
-    }
-
-    @Async
-    @Retryable(maxAttemptsExpression = "#{${letterAsync.maxAttempts}}", backoff = @Backoff(delayExpression = "#{${letterAsync.delay}}", multiplierExpression = "#{${letterAsync.multiplier}}", random = true))
+    @Retryable(maxAttemptsExpression = "#{@letterAsyncConfigProperties.maxAttempts}", backoff = @Backoff(delayExpression = "#{@letterAsyncConfigProperties.delay}", multiplierExpression = "#{@letterAsyncConfigProperties.multiplier}", random = true))
```
Comment (adusumillipraveen) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java:63]: This is unused.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java [hunk 2]
```diff
@@ -25,9 +27,24 @@ public SaveCorrespondenceAsyncService(CcdNotificationsPdfService ccdNotification
...
+    @Autowired
+    private LetterAsyncConfigProperties letterAsyncConfigProperties;
+
     @Async
-    @Retryable(maxAttemptsExpression = "#{${letterAsync.maxAttempts}}", backoff = @Backoff(delayExpression = "#{${letterAsync.delay}}", multiplierExpression = "#{${letterAsync.multiplier}}", random = true))
+    @Retryable(maxAttemptsExpression = "#{@letterAsyncConfigProperties.maxAttempts}", backoff = @Backoff(delayExpression = "#{@letterAsyncConfigProperties.delay}", multiplierExpression = "#{@letterAsyncConfigProperties.multiplier}", random = true, maxDelayExpression = "#{@letterAsyncConfigProperties.maxDelay}"))
     public void saveLetter(NotificationClient client, String notificationId, Correspondence correspondence, String ccdCaseId) throws NotificationClientException {
+
+        RetryContext context = RetrySynchronizationManager.getContext();
+        if (context != null && context.getRetryCount() == 0) {
```
Comment (adusumillipraveen) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java:38]: Making sure custom delay is added for on first try as 99% of first attempts fail.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java [hunk 3]
```diff
@@ -25,9 +27,24 @@ public SaveCorrespondenceAsyncService(CcdNotificationsPdfService ccdNotification
         this.ccdNotificationsPdfService = ccdNotificationsPdfService;
     }
 
+    @Autowired
+    private LetterAsyncConfigProperties letterAsyncConfigProperties;
+
     @Async
-    @Retryable(maxAttemptsExpression = "#{${letterAsync.maxAttempts}}", backoff = @Backoff(delayExpression = "#{${letterAsync.delay}}", multiplierExpression = "#{${letterAsync.multiplier}}", random = true))
+    @Retryable(maxAttemptsExpression = "#{@letterAsyncConfigProperties.maxAttempts}", backoff = @Backoff(delayExpression = "#{@letterAsyncConfigProperties.delay}", multiplierExpression = "#{@letterAsyncConfigProperties.multiplier}", random = true, maxDelayExpression = "#{@letterAsyncConfigProperties.maxDelay}"))
```
Comment (adusumillipraveen) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java:34]: Sonar was complaining about this format so had to change it. https://sonarsource.atlassian.net/browse/SONARJAVA-4917

File: src/main/resources/application.yaml [hunk 4]
```diff
@@ -1447,10 +1447,12 @@ feature.postHearings.enabled: ${POST_HEARINGS_FEATURE:false}
 feature.postHearingsB.enabled: ${POST_HEARINGS_B_FEATURE:false}
 feature.elinksV2.enabled: false
 
-letterAsync:
-  maxAttempts: ${LETTER_ASYNC_MAX_ATTEMPTS:5}
-  delay: ${LETTER_ASYNC_DELAY:6000}
-  multiplier: ${LETTER_ASYNC_MULTIPLIER:10}
+letter-async:
```
Comment (adusumillipraveen) [inline src/main/resources/application.yaml:1450]: Had to rename as per standards.

**Review by nitinprabhuhmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java (hunk 2 above)
Comment (Andrew-McMahon7) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java:38]: What happens if context is null, will it also try calling .getRetryCount and will this cause an error? Should the second part be a nested if statement?

Comment (adusumillipraveen) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/SaveCorrespondenceAsyncService.java:38]: No, Java will ignore later statement as soon as first statement is `false` as `false && (true/false)` will be `false`

File: src/main/resources/application.yaml [hunk 5]
```diff
@@ -1447,10 +1447,12 @@ feature.postHearings.enabled: ${POST_HEARINGS_FEATURE:false}
...
+  maxAttempts: ${LETTER_ASYNC_MAX_ATTEMPTS:10}
+  delay: ${LETTER_ASYNC_DELAY:10000}
+  multiplier: ${LETTER_ASYNC_MULTIPLIER:2}
+  maxDelay: ${LETTER_ASYNC_MAX_DELAY:180000}
+  initialDelay: ${LETTER_ASYNC_INITIAL_DELAY:5000}
```
Comment (Andrew-McMahon7) [inline src/main/resources/application.yaml:1455]: @adusumillipraveen do we need a separate pr to change these values in flux?

Comment (adusumillipraveen) [reply src/main/resources/application.yaml:1455]: No, the defaults are sensible as per our discussion. I checked that they are not set in flux, we can override if need be.

**Review by Andrew-McMahon7 — APPROVED** [review]

**Review by SomeshRA — APPROVED** [review]
